tests/integration/azure/test_azure_credentials.py

```diff
@@ -0,0 +1,142 @@
+        directory_id = end_point.split("/")[3]
+
+        name = f"testinfra_storageaccess_{make_random(4).lower()}"
+
+        azure_resource_permissions = MagicMock()
+        azure_resource_permissions.load.return_value = [
+            StoragePermissionMapping(
+                prefix="abfss://things@labsazurethings.dfs.core.windows.net/avoid_ext_loc_overlap",
+                client_id=application_id,
+                principal=name,
+                privilege="READ_FILES" if read_only else "WRITE_FILES",
+                directory_id=directory_id,
+            )
+        ]
+
+        azure_sp_crawler = MagicMock()
+        azure_sp_crawler.snapshot.return_value = [
+            AzureServicePrincipalInfo(
+                application_id=application_id,
+                secret_scope=secret_scope,
+                secret_key=secret_key,
+                tenant_id="test",
+                storage_account="test",
+            )
+        ]
+
+        installation = MagicMock()
+        installation.save.return_value = "azure_service_principal_migration_result.csv"
+
+        return {
+            "storage_credential_name": name,
+            "application_id": application_id,
+            "directory_id": directory_id,
+            "client_secret": client_secret,
+            "azure_resource_permissions": azure_resource_permissions,
+            "azure_sp_crawler": azure_sp_crawler,
+            "installation": installation,
+        }
+
+    return inner
+
+
+@pytest.fixture
+def execute_migration(ws):
+    def inner(variables: dict, integration_test_flag: str) -> AzureServicePrincipalMigration:
+        spn_migration = AzureServicePrincipalMigration(
+            variables["installation"],
+            ws,
+            variables["azure_resource_permissions"],
+            variables["azure_sp_crawler"],
+            integration_test_flag=integration_test_flag,
+        )
+        spn_migration.execute_migration(
+            MockPrompts({"Above Azure Service Principals will be migrated to UC storage credentials *": "Yes"})
+        )
+        return spn_migration
+
+    return inner
+
+
+def test_spn_migration_existed_storage_credential(
+    ws, execute_migration, make_storage_credential_from_spn, prepare_spn_migration_test
+):
+    variables = prepare_spn_migration_test(read_only=False)
+
+    # create a storage credential for this test
+    make_storage_credential_from_spn(
+        name=variables["storage_credential_name"],
+        application_id=variables["application_id"],
+        client_secret=variables["client_secret"],
+        directory_id=variables["directory_id"],
+    )
+
+    # test that the spn migration will be skipped due to above storage credential is existed
+    spn_migration = execute_migration(variables, integration_test_flag=variables["storage_credential_name"])
+
+    # because storage_credential is existing, no spn should be migrated
+    assert not spn_migration._final_sp_list
```

Comment:
using private methods is not allowed outside of classes that they're declared in. this is going to break.

Reply:
fixed